optuna/study.py

```diff
@@ -594,7 +594,12 @@ def _optimize_sequential(
                 if elapsed_seconds >= timeout:
                     break
 
-            self._run_trial_and_callbacks(func, catch, callbacks, gc_after_trial)
+            try:
+                self._run_trial_and_callbacks(func, catch, callbacks, gc_after_trial)
+            except exceptions.StopStudy:
```

Comment:
Current implementation seems to catch `StopStudy` only when it’s raised in the callback. How about also covering those raised in the objective function?

Reply:
What status should we assign to a trial, when `StopStudy` is raised in the objective function? I think `optuna.structs.TrialState.COMPLETE` is a reasonable choice, but there is no value we can assign to the trial in this case. I think this is an issue, so may be we should avoid raising `StopStudy` inside objective function.

Reply:
Good catch. The expected state is not trivial.

On the other hand, it might be a burden for the user to understand that "a callback is acceptable but an objective function is not."